### PR TITLE
run sdpa with dtensor

### PR DESCRIPTION
**Pull request by tianyu-l**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #180
* #285
* #161
* #172

This PR gets rid of the manual adjustment of num of heads in attention layers, by using dtensor outputs of `wq`, `wk`, `wv`, so that the SDPA is aware of the distributedness.

### CODE REVIEW

File: torchtrain/parallelisms/parallelize_llama.py [hunk 1]
```diff
@@ -177,9 +184,9 @@ def parallelize_llama(model, world_mesh, parallel_dims, job_config: JobConfig):
                     input_layouts=(Shard(0), None),
                     desired_input_layouts=(Replicate(), None),
                 ),
-                "attention.wq": col_parallel_strategy(),
-                "attention.wk": col_parallel_strategy(),
-                "attention.wv": col_parallel_strategy(),
+                "attention.wq": col_parallel_strategy(use_local_output=False),
```
Comment (wanchaol) [inline torchtrain/parallelisms/parallelize_llama.py]: 🤔 I thought we need to replicate the `freq_cis` but here it seems we don't need to?

**Comment by tianyu-l** [comment]
> just curious, is this gonna land soon or does it have some risk or unfinished business?
> 
> also looks like this could use a rebase. i got a little confused applying it on my branch bc some of the sharding config seems changed (attention.wo and attention_norm)

It hasn't been landed because there is a very strange bug (#267) associated with (but seemingly not caused by) multiplication using DTensor. It would be triggered in the rotary embedding computation if this PR is landed. I will work on the bug soon since it will also benefit PP (iiuc). @wconstab

**Comment by wconstab** [comment]
> It would be triggered in the rotary embedding computation if this PR is landed

oh, is this related to dispatching for complex numbers by any chance?

**Comment by tianyu-l** [comment]
> oh, is this related to dispatching for complex numbers by any chance?

@wconstab Possibly, we don't know. The `aten.mul` op returns bad results with inputs being raw torch.Tensor (desugared from DTensor), and this bug is only present in the backward pass. Do you know who I should ask for help from?

**Pull request closed without merging**